src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -420,122 +382,127 @@ Batch::doApply()
+            preclaimResult.j);
+        auto const applyResult = invoke_apply(actx);
 
-    TER preResult = tesSUCCESS;
-    ApplyViewImpl& avi = dynamic_cast<ApplyViewImpl&>(ctx_.view());
-    for (auto const& dryRun : dryVector)
-    {
         STObject meta{sfBatchExecution};
-        meta.setFieldU8(sfTransactionResult, TERtoInt(dryRun.second));
-        meta.setFieldU16(sfTransactionType, dryRun.first);
+        meta.setFieldU8(sfTransactionResult, TERtoInt(applyResult.first));
+        meta.setFieldU16(sfTransactionType, stx.getTxnType());
+        if (applyResult.first == tesSUCCESS)
+            meta.setFieldH256(sfTransactionHash, stx.getTransactionID());
+
         avi.addBatchExecutionMetaData(std::move(meta));
 
-        // tfAllOrNothing
-        if (dryRun.second != tesSUCCESS && flags & tfAllOrNothing)
-        {
-            preResult = tecBATCH_FAILURE;
-        }
-    }
+        accountCount[stx.getAccountID(sfAccount)] += 1;
 
-    // WET RUN
-    TER result = tesSUCCESS;
-    if (preResult == tesSUCCESS)
-    {
-        std::vector<STObject> batch;
-        avi.setHookMetaData(std::move(batch));
-        for (std::size_t i = 0; i < stxTxns.size(); ++i)
+        if (applyResult.first != tesSUCCESS)
         {
-            auto const& stx = stxTxns[i];
-            ApplyContext actx(
-                ctx_.app,
-                ctx_.base_,
-                stx,
-                preclaimResults[i],
-                ctx_.view().fees().base,
-                ctx_.base_.open() == 1 ? tapPREFLIGHT_BATCH
-                                       : ctx_.view().flags(),
-                ctx_.journal);
-            auto const _result = invoke_apply(actx);
-
-            STObject meta{sfBatchExecution};
-            meta.setFieldU8(sfTransactionResult, TERtoInt(_result.first));
-            meta.setFieldU16(sfTransactionType, stx.getTxnType());
-            if (_result.first == tesSUCCESS)
-                meta.setFieldH256(sfTransactionHash, stx.getTransactionID());
-
-            avi.addBatchExecutionMetaData(std::move(meta));
-
-            if (_result.first != tesSUCCESS)
+            if (flags & tfUntilFailure)
             {
-                if (flags & tfUntilFailure)
-                {
-                    actx.discard();
-                    result = tesSUCCESS;
-                    break;
-                }
-                if (flags & tfOnlyOne)
-                {
-                    actx.discard();
-                    continue;
-                }
+                actx.discard();
+                result = tesSUCCESS;
+                break;
             }
-
-            if (_result.first == tesSUCCESS && flags & tfOnlyOne)
+            if (flags & tfOnlyOne)
             {
+                actx.discard();
+                continue;
+            }
+            if (flags & tfAllOrNothing)
+            {
+                accountCount.clear();
+                ctx_.base_.rawRevert();
```

Comment:
I haven't deeply looked at this code, so I am not up on what has changed so far, but I really do not like this `rawRevert` function. The canonical way to reset / revert the view during transaction processing is `ctx_.discard()`. `base_` is currently not touched until the transactor is fully complete and `ctx_.apply()` is called. In fact, `base_` is private in the current `develop`, and I don't see any need to change that.

Because this is `Batch`, this transactor is invoking other `Transactor` classes. That's obviously necessary. My guess is that you're reusing the `ctx_` from this object when invoking those other objects. Thus, when they finish, they call `apply()` and modify `base_`.

There's probably a better way. It comes down to this constructor: https://github.com/XRPLF/rippled/blob/develop/src/xrpld/ledger/OpenView.h#L171

So in `Batch::doApply`, I think it would be really good to do something like this:
```
    [...]
    Sandbox sb(ctx_.view());
    bool changed = false;
    auto const& txns = ctx_.tx.getFieldArray(sfRawTransactions);
    for (STObject& txn : txns)
    {
        OpenView subView(sb);
        auto const [ter, applied] = ripple::apply(
            ctx_.app, subView, tx, tapPREFLIGHT_BATCH, ctx_.journal);
        [... Do whatever with `ter` ...]
        if (applied)
        {
            subView.apply(sb, tx, result.first, journal);
            changed = true;
        }
    }
    [...]
    if (changed)
    {
        sb.apply(ctx_.rawView());
    }
    [...]
```
This code might actually compile as-is (other than the `[...]`s of course), but no promises.

Reply:
@ximinez Thank you for your review. I agree with you. Will update.